lighthouse-core/audits/resource-budget.js

```diff
@@ -0,0 +1,182 @@
+/**
+ * @license Copyright 2019 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+const Audit = require('./audit.js');
+const ResourceSummary = require('../computed/resource-summary.js');
+const i18n = require('../lib/i18n/i18n.js');
+
+const UIStrings = {
+  /** Imperative title of a Lighthouse audit that tells the user to minimize the size and quantity of resources used to load the page. */
+  title: 'Keep request counts low and file sizes small',
+  /** Description of a Lighthouse audit that tells the user that they can setup a budgets for the quantity and size of page resources. No character length limits. */
+  description: 'To set budgets for the quantity and size of page resources,' +
+    ' add a budget.json file.',
+  /** [ICU Syntax] Label identifying the number of requests*/
+  requestCount: `{count, plural,
+    =1 {1 request}
+    other {# requests}
+   }`,
+};
+
+const str_ = i18n.createMessageInstanceIdFn(__filename, UIStrings);
+
+/** @typedef {{count: number, size: number}} ResourceEntry */
+
+class ResourceBudget extends Audit {
+  /**
+   * @return {LH.Audit.Meta}
+   */
+  static get meta() {
+    return {
+      id: 'resource-budget',
+      title: str_(UIStrings.title),
+      description: str_(UIStrings.description),
+      scoreDisplayMode: Audit.SCORING_MODES.INFORMATIVE,
+      requiredArtifacts: ['devtoolsLogs', 'URL'],
+    };
+  }
+
+  /**
+   * @param {LH.Budget | undefined} budget
+   * @return {LH.Audit.Details.Table['headings']}
+   */
+  static tableHeadings(budget) {
+    /** @type {LH.Audit.Details.Table['headings']} */
+    const headers = [
+      {key: 'label', itemType: 'text', text: 'Resource Type'},
+      {key: 'requestCount', itemType: 'numeric', text: 'Requests'},
+      {key: 'size', itemType: 'bytes', text: 'Transfer Size'},
+    ];
+
+    /** @type {LH.Audit.Details.Table['headings']} */
+    const budgetHeaders = [
+      {key: 'countOverBudget', itemType: 'text', text: ''},
+      {key: 'sizeOverBudget', itemType: 'bytes', text: 'Over Budget'},
+    ];
+    return budget ? headers.concat(budgetHeaders) : headers;
+  }
+
+  /**
+   * @param {LH.Budget.ResourceType} resourceType
+   * @return {string}
+   */
+  static rowLabel(resourceType) {
+    /** @type {Record<LH.Budget.ResourceType,string>} */
+    const strMappings = {
+      'total': str_(i18n.UIStrings.totalResourceType),
```

Comment:
`str_()` is pretty fast but it does do sorting out the right string to reference...rather than call it for all the resourceType strings * number of times `rowLabel` is called, maybe make this a map to the raw UIStrings e.g. `'total': i18n.UIStrings.totalResourceType,`, then down on L115 do `const label = str_(this.getRowLabel(type));`